File: HDOJ/CCPC16Online/src/1010.cc
/*************************************************************************
  >         File: 1010.cc
  >       Author: Riho.Yoshioka
  >         Mail: riho.yoshioka@yandex.com
  > Created Time: Sun 14 Aug 2016 03:52:42 PM CST
*************************************************************************/
#include <bits/stdc++.h>
typedef long long LL;
using namespace std;

const int N = 1e5 + 5;
const int K = 20;
const int L = 16;

int tr[N * K * K][2], node;
int rt[N], idx[N], res[N], ans[N];
vector<int> numbers[N];

int first[N], to[N << 1], nex[N << 1], edge;

void Add(int x, int y) {
  nex[++edge] = first[x];
  first[x] = edge;
  to[edge] = y;
}

int Node() {
  tr[node][0] = tr[node][1] = 0;
  return node++;
}

void Init(int n) {
  node = 0;
  Node();
  for (int i = 1; i <= n; i++) rt[i] = Node();
}

void Insert(int a, int x) {
  int now = rt[a];
  numbers[a].push_back(x);
  for (int i = L; i >= 0; i--) {
    int idx = (x & (1 << i)) > 0;
    if (tr[now][idx] == 0) tr[now][idx] = Node();
    now = tr[now][idx];
  }
}

int Query(int a, int x) {
  int now = rt[a];
  int number = 0;
  for (int i = L; i >= 0; i--) {
    int idx = (x & (1 << i)) > 0;
    if (tr[now][idx] == 0) idx ^= 1;
    number |= idx << i;
    now = tr[now][idx];
  }
  return x ^ number;
}

int Merge(int a, int b) {
  a = idx[a], b = idx[b];
  if (numbers[a].size() < numbers[b].size()) swap(a, b);
  int ans = max(res[a], res[b]);
  for (auto &y: numbers[b]) {
    ans = max(ans, Query(a, y));
    Insert(a, y);
  }
  res[a] = ans;
  return a;
}

void dfs(int x, int p) {
  for (int go = first[x]; go; go = nex[go]) {
    int y = to[go];
    if (y == p) continue;
    dfs(y, x);
    idx[x] = Merge(x, y);
  }
  ans[x] = numbers[idx[x]].size() > 1 ? res[idx[x]] : -1;
}

int in() {
  char c;
  while (c = getchar(), (c < '0' || c > '9') && (c != '-'));
  bool flag = (c == '-');
  if (flag) c = getchar();
  int x = 0;
  while (c >= '0' && c <= '9') {
    x = x * 10 + c - 48;
    c = getchar();
  }
  return flag ? -x : x;
}

void out(int x) { //int
  if (x < 0) putchar('-'), x = -x;
  int len = 0, bit[10]; // LL  -> bit[20]
  while (x) {
    bit[len++] = x % 10;
    x /= 10;
  }
  if (!len) bit[len++] = 0;
  while (len--) putchar(bit[len] + 48);
  putchar('\n');
}

int main() {
  int t, ca = 1;
  t = in();
  //scanf("%d", &t);
  while (t--) {
    int n;
    n = in();
    //scanf("%d", &n);
    Init(n);
    edge = 0;
    for (int i = 1; i <= n; i++) {
      int x;
      x = in();
      //scanf("%d", &x);
      idx[i] = i;
      res[i] = 0;
      first[i] = 0;
      numbers[i].clear();
      Insert(idx[i], x);
    }
    for (int i = 0; i < n - 1; i++) {
      int x, y;
      x = in(), y = in();
      //scanf("%d%d", &x, &y);
      Add(x, y);
      Add(y, x);
    }
    dfs(1, 0);
    int q;
    q = in();
    //scanf("%d", &q);
    printf("Case #%d:\n", ca++);
    while (q--) {
      int x;
      x = in();
      //scanf("%d", &x);
      out(ans[x]);
      //printf("%d\n", ans[x]);
    }
  }
  return 0;
}

File: HDOJ/CCPC16Online/src/1010.java
/*************************************************************************
  >         File: Main.java
  >       Author: Riho.Yoshioka
  >         Mail: riho.yoshioka@yandex.com
  > Created Time: Sun 14 Aug 2016 03:04:45 PM CST
*************************************************************************/
import java.io.*;
import java.util.*;

public class Main {
  static final int K = 2;
  public static void main(String[] args) {
    InputStream inputStream = System.in;
    OutputStream outputStream = System.out;
    InputReader in = new InputReader(inputStream);
    PrintWriter out = new PrintWriter(outputStream);
    
    TaskA solver = new TaskA();
    solver.solve(1, in, out);
    out.close();
  }

  static class TaskA {
    ArrayList<Integer>[] edges;
    int[] ans;
    Trie[] tr;
    public void solve(int testNumber, InputReader in, PrintWriter out) {
      int t = in.nextInt();
      for (int ca = 1; ca <= t; ca++) {
        int n = in.nextInt();
        edges = new ArrayList[n + 1];
        tr = new Trie[n + 1];
        ans = new int[n + 1];
        for (int i = 1; i <= n; i++) {
          int x = in.nextInt();
          edges[i] = new ArrayList<>();
          tr[i] = new Trie();
          tr[i].insert(x);
        }
        for (int i = 0; i < n - 1; i++) {
          int x, y;
          x = in.nextInt(); y = in.nextInt();
          edges[x].add(y);
          edges[y].add(x);
        }
        dfs(1, 0);
        int q = in.nextInt();
        out.println("Case #" + ca + ":");
        for (int i = 0; i < q; i++) {
          int idx = in.nextInt();
          out.println(ans[idx]);
        }
      }
    }

    private void dfs(int x, int p) {
      for (int y: edges[x]) {
        if (y == p) continue;
        dfs(y, x);
        assert tr[x] != null && tr[y] != null;
        tr[x] = tr[x].merge(tr[x], tr[y]);
      }
      ans[x] = tr[x].numbers.size() < 2 ? -1 : tr[x].ans;
    }
  }

  static class Node {
    Node[] son;

    Node () {
      son = new Node[2];
    }
  }

  static class Trie {
    ArrayList<Integer> numbers;
    int ans;
    Node root;

    Trie () {
      root = new Node();
      numbers = new ArrayList<>();
      ans = 0;
    }

    void insert(int x) {
      Node now = root;
      numbers.add(x);
      for (int i = K; i >= 0; i--) {
        int idx = (x & (1 << i)) != 0 ? 1 : 0;
        if (now.son[idx] == null) now.son[idx] = new Node();
        now = now.son[idx];
      }
    }

    int query(int x) {
      Node now = root;
      int number = 0;
      assert now != null;
      for (int i = K; i >= 0; i--) {
        int idx = (x & (1 << i)) != 0 ? 1 : 0;
        if (now.son[idx] == null) idx ^= 1;
        number |= (idx << i);
        now = now.son[idx];
      }
      return number ^ x;
    }

    Trie merge(Trie a, Trie b) {
      if (a.numbers.size() < b.numbers.size()) {
        Trie tmp = a; a = b; b = tmp;
      }
      Trie res = a;
      res.ans = Math.max(a.ans, b.ans);
      for (int x: b.numbers) {
        int now = res.query(x);
        res.ans = Math.max(ans, now);
        res.insert(x);
      }
      return res;
    }
  }

  static class InputReader {
    public BufferedReader reader;
    public StringTokenizer tokenizer;

    public InputReader(InputStream stream) {
      reader = new BufferedReader(new InputStreamReader(stream), 32768);
      tokenizer = null;
    }

    public String next() {
      while (tokenizer == null || !tokenizer.hasMoreTokens()) {
        try {
          tokenizer = new StringTokenizer(reader.readLine());
        } catch (IOException e) {
          throw new RuntimeException(e);
        }
      }
      return tokenizer.nextToken();
    }

    public int nextInt() {
      return Integer.parseInt(next());
    }

    public long nextLong() {
      return Long.parseLong(next());
    }

    public double nextDouble() {
      return Double.parseDouble(next());
    }
  }
}
